ext/xbrz/xbrz_tools.h

```diff
@@ -0,0 +1,278 @@
+// ****************************************************************************
+// * This file is part of the xBRZ project. It is distributed under           *
+// * GNU General Public License: https://www.gnu.org/licenses/gpl-3.0         *
+// * Copyright (C) Zenju (zenju AT gmx DOT de) - All Rights Reserved          *
+// *                                                                          *
+// * Additionally and as a special exception, the author gives permission     *
+// * to link the code of this program with the following libraries            *
+// * (or with modified versions that use the same licenses), and distribute   *
+// * linked combinations including the two: MAME, FreeFileSync, Snes9x, ePSXe *
+// * You must obey the GNU General Public License in all respects for all of  *
+// * the code used other than MAME, FreeFileSync, Snes9x, ePSXe.              *
+// * If you modify this file, you may extend this exception to your version   *
+// * of the file, but you are not obligated to do so. If you do not wish to   *
+// * do so, delete this exception statement from your version.                *
+// ****************************************************************************
+
+#ifndef XBRZ_TOOLS_H_825480175091875
+#define XBRZ_TOOLS_H_825480175091875
+
+#include <cassert>
+#include <algorithm>
+#include <type_traits>
+
+
+namespace xbrz
+{
+template <uint32_t N> inline
+unsigned char getByte(uint32_t val) { return static_cast<unsigned char>((val >> (8 * N)) & 0xff); }
+
+inline unsigned char getAlpha(uint32_t pix) { return getByte<3>(pix); }
+inline unsigned char getRed  (uint32_t pix) { return getByte<2>(pix); }
+inline unsigned char getGreen(uint32_t pix) { return getByte<1>(pix); }
+inline unsigned char getBlue (uint32_t pix) { return getByte<0>(pix); }
```

Comment:
The order of these components is wrong - you might've noticed it looking funny in games?  See the old code which Durante altered when adding xBRZ support to PPSSPP:

```c++
inline unsigned char getRed  (uint32_t pix) { return getByte<0>(pix); }
inline unsigned char getGreen(uint32_t pix) { return getByte<1>(pix); }
inline unsigned char getBlue (uint32_t pix) { return getByte<2>(pix); }
inline unsigned char getAlpha(uint32_t pix) { return getByte<3>(pix); }
```

Do graphics look right generally otherwise?  Definitely best to make sure each backend's colors are right and not RGBA swizzled or crashing due to wrong buffer size or something.  Those are the most likely problems.

-[Unknown]

Reply:
I have only tested this in Driver 76 and I have not observed any obvious miscolorations (and I'm pretty sensitive to spotting such). If anything, that would mean we are using a wrong xBRZ mode? IIRC they allow for ARGB and RGBA modes - the current code clearly uses ARGB while Durante's changes seem to make it use ABGR/RGBA.

So if anything, modes should be addressed instead of swizzling channels in the code?

Reply:
Hm, I dunno then.  I remember on previous updates it caused problems and not in all games, but I don't remember why.  But again, I'm pretty sure it depends on backend.

If we're swizzling channels on the CPU as a first pass, it probably means we're making this CPU-based scaling even slower.  Hopefully we are not doing that.

-[Unknown]

Reply:
So far I only tested with D3D11. I'll check all the backends later and maybe against more games than just Driver 76.